src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1402 @@
+
+    def generate_text_semantic(
+        self,
+        input_ids: torch.Tensor,
+        history_prompt: Optional[Dict[str, torch.Tensor]] = None,
+        semantic_generation_config: BarkSemanticGenerationConfig = None,
+        **kwargs,
+    ) -> torch.LongTensor:
+        if semantic_generation_config is None:
+            # TODO (joao): workaround until nested generation config is compatible with PreTrained Model
+            semantic_generation_config = BarkSemanticGenerationConfig(**self.generation_config.semantic_config)
+
+        batch_size = input_ids.shape[0]
+
+        max_input_semantic_length = semantic_generation_config.max_input_semantic_length
+
+        input_ids = input_ids + semantic_generation_config.text_encoding_offset
+
+        if kwargs.get("attention_mask", None) is not None:
+            input_ids.masked_fill_(
+                (1 - kwargs.pop("attention_mask")).bool(), semantic_generation_config.text_pad_token
+            )
+
+        if history_prompt is not None:
+            semantic_history = history_prompt["semantic_prompt"][
+                -semantic_generation_config.max_input_semantic_length :
+            ]
+            semantic_history = nn.functional.pad(
+                semantic_history,
+                (0, max_input_semantic_length - len(semantic_history)),
+                value=semantic_generation_config.semantic_pad_token,
+                mode="constant",
+            )
+        else:
+            semantic_history = torch.tensor(
+                [semantic_generation_config.semantic_pad_token] * max_input_semantic_length, dtype=torch.int
+            ).to(self.device)
+
+        semantic_history = torch.repeat_interleave(semantic_history[None], batch_size, dim=0)
+
+        infer_array = torch.tensor(
+            [[semantic_generation_config.semantic_infer_token]] * batch_size, dtype=torch.int
+        ).to(self.device)
+
+        input_embeds = torch.cat(
+            [
+                self.semantic.wte(input_ids[:, :max_input_semantic_length])
+                + self.semantic.wte(semantic_history[:, : max_input_semantic_length + 1]),
+                self.semantic.wte(infer_array),
+            ],
+            dim=1,
+        )
+
+        semantic_logits_processor = SemanticLogitsProcessor(
+            semantic_generation_config.semantic_vocab_size, semantic_generation_config.semantic_pad_token
+        )
+
+        # TODO: for now, it is not implemented yet as long as StoppingCriteria issue is not dealt with
+        # https://github.com/huggingface/transformers/issues/23674
+        # semantic_stopping_criteria = SemanticStoppingCriteria(min_eos_p , semantic_generation_config.semantic_pad_token)
+        # TODO: add a max_gen_duration_s early stop
+
+        # pass input_ids in order to stay consistent with the transformers generate method even though it is not used (except to get the input seq_len - that's why we keep the first 257 tokens)
```

Comment:
Lot's of commented code here - can you pick out what we need and what's still missing, and address?